Thread interrupt cleanup

lhelwerd opened the pull request:
- Do not interrupt the main thread from unregistered threads. This should stop tests from failing when they actually succeed, but have threads running that have not yet detected that they are already deactivated.
- Clean up "None" spam in logs when a script finished normally.
- Display logs in the test runner before clearing them.

This should close #75 but we should probably make sure that all intermittent failures are now gone. @timvandermeij can you still reproduce the mission_cycle test failure mentioned in that bug?

lhelwerd committed:
Thread_Manager: Do not interrupt main thread from unregistered threads

When a ThreadtableTestCase is finished, we destroy the thread manager.
This deactivates all the threads and unregisters them. However, some
threads may be blocking/unresponsive for some time, causing them to
check that they are deactivated much later. Even worse, they might hit
an error, such as a closed connection, in between. As normal, the error
interrupts the main thread, which might be performing another test or
have exited already. This respectively cause a KeyboardInterrupt or
segmentation fault.

No longer interrupt the main thread for threads that are (already)
unregistered. Add tests for this, and add a test for the log method as
an extra.

lhelwerd committed:
Thread_Manager: Do not log destroy when not in exception context

To prevent spam of "None" errors in the logs, skip the log call in
destroy when there is no info about the exception. Since we use destroy
when ending programs normally as well, the logs become too cluttered
with non-errors due to this. The interrupt method will still
unconditionally log, since it has useful information about the source of
the call and is specifically targeted for exception handling.

lhelwerd committed:
Test runner: Display logs before clearing them

In the future, to be able to find out why tests fail intermittently, the
logs may provide useful details about exceptions that are handled by the
interpreter but still cause spurious errors. These should now be mostly
resolved by the Thread_Manager changes. The logs are now at least useful
to print, so this seems like a helpful idea.

lhelwerd committed:
Thread_Manager: Clean up naming

The thread manager uses "thread" to refer to Threadable objects
registered in it, but since we now import the thread module this causes
confusing name clashing. Python knows just fine how to handle it, but
for clarity we rename the thread variables to threadable.

timvandermeij commented:
I'm now getting this in the console for the `mission_cycle` test (sometimes):

<pre>
2016-03-26 20:07:55,377 - ERROR - 'line_follower' thread
Traceback (most recent call last):
  File "/home/timvandermeij/Documenten/Universiteit/5/Researchproject en masterscriptie/Code/mobile-radio-tomography/location/Line_Follower.py", line 51, in _loop
    sensor_values = self.read()
  File "/home/timvandermeij/Documenten/Universiteit/5/Researchproject en masterscriptie/Code/mobile-radio-tomography/location/Line_Follower_Arduino.py", line 54, in read
    line = self._serial_connection.readline()
  File "/home/timvandermeij/.local/lib/python2.7/site-packages/serial/serialposix.py", line 511, in read
    raise SerialException('read failed: %s' % (e,))
SerialException: read failed: (9, 'Ongeldige bestandsdescriptor')
</pre>

timvandermeij commented:
There are more failures now uncovered:

<pre>
======================================================================
ERROR: test_log (mobile-radio-tomography.tests.core_thread_manager.TestCoreThreadManager)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/timvandermeij/Documenten/Universiteit/5/Researchproject en masterscriptie/Code/mobile-radio-tomography/tests/core_thread_manager.py", line 118, in test_log
    self.assertEqual(self.thread_manager._logger, logger_mock)
AttributeError: 'Thread_Manager' object has no attribute '_logger'

======================================================================
FAIL: test_step (mobile-radio-tomography.tests.mission_cycle.TestMissionCycle)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/timvandermeij/Documenten/Universiteit/5/Researchproject en masterscriptie/Code/mobile-radio-tomography/tests/mission_cycle.py", line 80, in test_step
    self.assertEqual(self.vehicle._current_waypoint, 0)
AssertionError: -1 != 0
</pre>

timvandermeij commented:
I did some more research and found the following:

- The `mission_cycle` error still occurs (I have seen it three times now).
- The serial exception I reported above is because of the unconditional line reading in https://github.com/timvandermeij/mobile-radio-tomography/blob/master/location/Line_Follower_Arduino.py#L54. Apparently this does not play well with the PTY. I tried moving it into the `try` block and that fixes the exception, but generally makes the unit tests much slower because it will constantly get in the loop at https://github.com/timvandermeij/mobile-radio-tomography/blob/master/location/Line_Follower_Arduino.py#L53 until the unit test framework kills it. I also tried catching the serial exception there and returning, but since other components and tests expect that this method always returns a valid list we would not some adjustments there too. That is also not desired because it requires converting an exception to a silent ignore only for the unit tests, because usually we want an exception to be raised. I'm not sure what to do with all this... Ideas?

timvandermeij committed:
Unit tests: actually verify that the logger is lazily initialized

The test only checked the state after initialization, but it did not
test if the logger is indeed uninitialized before the call to the
logging method.

lhelwerd commented:
I'm thinking the cause lies more in the unit test framework itself, since I have not yet seen this on the actual hardware when running the actual code.

My target is the code starting from https://github.com/timvandermeij/mobile-radio-tomography/blob/master/tests/core_usb_manager.py#L21 where we create virtual serial ports for the USB manager. There could be multiple problems here:
- We are opening a master and a slave, where the "master" is meant to be "our" communication end while the "slave" is meant to be the device; compare with [Wikipedia's definition](https://en.wikipedia.org/wiki/Pseudoterminal) which is about terminals specifically. However, the `DEVNAME`s are set to the slave port, which is the wrong end for our purposes. We should pass the master end here instead, I believe.
- Due to the confusion above, we _only_ close the slave end _twice_ in `tearDown`, once by `clear` and once by `os.close`. If we would use the master port for the `DEVNAME`s then this would be solved as well, since then `clear` closes the master end and the test closes the slave end.
- We supply the same PTY to the XBee device, serial device and other device. That could potentially cause data being read/written twice or a port being closed while the other is still rightfully using it.

timvandermeij committed:
Unit tests: comment improvements

Furthermore we're consistent with newlines after comment blocks in
`core/Threadable.py`.

timvandermeij commented:
Actually that sounds like a good point. I'll check what we can do here.

timvandermeij commented:
I have now made the ports for the XBee, TTL and other devices different. I noticed that the slave is actually the `/dev/pty/*` device, so we should pass the slave as the device name to the USB manager. Components will then write to the slave and we can read this in the unit tests by checking the master, like we already do. It does not work when we reverse the master and slave because the master is always the same device, namely `/dev/ptmx`, and problems are reported when writing to that. If we were to pass that, we would also still be writing to the same device. Therefore it seems like it's fine this way: the components write to the slave and we read from the master.

This is a nice improvement, but it does not fix any of the intermittent issues. I did however find the cause of the intermittent serial issues. There is a race condition between the line follower Arduino and the USB testcase. The line follower Arduino constantly reads from the serial connection without a timeout (for some reason, I presume it's for a fast throughput of data). Due to the multiple interitance of the test case classes, sometimes it happens that the USB manager kills the PTY devices before the line follower Arduino thread is killed. The line follower Arduino loop will attempt to read from the closed device, resulting in all kinds of serial errors (I have seen errors about incorrect device descriptors, incorrectly parsed data and missing devices). Unfortunately, I cannot think of a way to ensure that the line follower Arduino thread is killed before we kill the PTY devices. Maybe you have an idea? Maybe we can think of an alternative to the PTY devices (such as mocking it in some way)?

I'm still not sure why the `mission_cycle` test keeps failing intermittently. It does not seem to be related to the serial errors, so I think somehow the test is faulty or there is another race condition we don't know of there.

I'll push my fixes now since they make sense, but we should probably figure out the issues above before merging this.

timvandermeij committed:
Unit tests: refactor PTY usage in the USB test case class

Use separate PTYs for the XBee, TTL and other devices to avoid
conflicts. Furthermore, use the device's `close()` method instead of
`os.close()` for safety.

lhelwerd commented:
Perhaps you should open the master side as `os.fdopen(master_ttl, 'r+')` so that you can actually read _and_ write to it? The way things currently are seem backward to me. Note that the `/dev/ptmx` has some trickery so that it is not the same device for different pty numbers; that is why you have to open it using the file descriptor number with `os.fdopen`, not the ttyname. Also, the master and slave are not the same device (they just have their read and write ports intertwined with each other), and not all components only write; the line follower being a notable exception.

I can only reproduce the "bad file descriptor" exception when I open the same pty twice. That is why I hoped the last solution would at least fix those errors. Maybe there are other places that open them twice?

Otherwise, we can mock the `Line_Follower_Arduino.read` method or the entire `Line_Follower._loop` in the `mission_cycle` test. If it is the only cause of the intermittent failures, then that might be the best solution. The test does not really test the line follower anyway; it does not have a good way to do so and we have other tests for that.

The serial connection in the `read` method probably was probably blocking indefinitely then, since there will never be any data on `readline`. Or perhaps at some point there is, because the pseudopty seems to exhibit strange behavior on some OSes...

As implied in #75, I cannot reproduce these intermittent errors, which makes it difficult to debug. Perhaps you can put some prints in `Line_Follower_Arduino.read` to see what it is doing, and when? Because if it does read data, then the robot vehicle might receive a callback that causes it to get in the wrong state, thus failing the test.

timvandermeij commented:
Good points, I'll look into it first thing tomorrow. Thanks!

lhelwerd commented:
Given that we supply device names and not open file descriptors to the USB manager test, I suppose we have to live with the device being the master and us reading/writing to the slave. The backwardness of it all does not call for such a rewrite of the USB manager interface to allow dropping in open files into it.

Unless... An alternative option for testing the actual serial interface is to use sockets. pySerial supports opening sockets as serial interfaces using [serial_for_url](http://pythonhosted.org/pyserial/pyserial_api.html#serial.serial_for_url). Passing it a [socket URL](http://pythonhosted.org/pyserial/url_handlers.html) makes it open one socket. Then we can bind a normal socket to it and have more control over the tests. `serial_for_url` also supports normal device names as usual, so we should be able to convert the USB manager to use it.

I'm still unsure whether and to what extend it will nullify the intermittent errors, and it's such a specific test-only problem that we should not spend all our time fixing it.

(I would, at least, recommend to have a good breakfast first thing tomorrow. :egg:)

timvandermeij committed:
Unit tests: patch the `Line_Follower._loop()` method

We are not actually testing the line follower in the test anyway and it
is causing intermittent failures.

timvandermeij commented:
I decided to patch the `Line_Follower._loop()` method in the `mission_cycle` unit test. It is the easiest solution (because I also don't want to change a lot of code just for the tests) and in my testing it seems to fix everything. I ran the unit tests 20 times locally and all of them passed without any errors (i.e., the serial exceptions are resolved and the mission cycle 0 != -1 failure is also resolved). I'll do several runs (10 or so) on Travis CI too and if there are no further issues, I'll review and merge this PR and consider the intermittent issues as fixed.

timvandermeij commented:
In summary, after 20 local test runs on 10 test runs on Travis CI there are no more intermittent errors. The segmentation faults, serial exceptions and `mission_cycle` failures are no longer reproduced. Let's merge this!
